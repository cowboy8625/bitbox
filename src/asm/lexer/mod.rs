use std::{iter::Peekable, str::Chars};

pub fn lex(src: &str) -> Vec<Token> {
    Lexer::new(src).lex()
}

#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
pub struct Span {
    pub row_start: usize,
    pub row_end: usize,
    pub col_start: usize,
    pub col_end: usize,
    pub byte_start: usize,
    pub byte_end: usize,
}

impl From<(Span, Span)> for Span {
    fn from((start, end): (Span, Span)) -> Self {
        Self {
            row_start: start.row_start,
            row_end: end.row_end,
            col_start: start.col_start,
            col_end: end.col_end,
            byte_start: start.byte_start,
            byte_end: end.byte_end,
        }
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    KeywordPush,
    KeywordPop,
    KeywordLoad,
    KeywordAdd,
    KeywordSub,
    KeywordDiv,
    KeywordMul,
    KeywordEq,
    KeywordJne,
    KeywordInc,
    KeywordHult,
    KeywordPrintReg,
    KeywordAnd,
    KeywordOr,
    Number(u32),
    Identifier(String),
    Colon,
    Comma,
    Period,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    PercentSign,
    Ampersand,
    Equal,
    Delimiter,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

struct Lexer<'a> {
    src: Peekable<Chars<'a>>,
    span: Span,
    tokens: Vec<Token>,
    current: Option<char>,
    last_char: Option<char>,
}

impl<'a> Lexer<'a> {
    fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            span: Span::default(),
            tokens: Vec::new(),
            current: None,
            last_char: None,
        }
    }

    fn next(&mut self) -> Option<char> {
        self.last_char = self.current;
        self.current = self.src.next();
        match self.current {
            Some(c) if matches!(self.last_char, Some('\n')) => {
                self.span.row_end += 1;
                self.span.col_start = 0;
                self.span.col_end = 1;
                self.span.byte_end += 1;
                Some(c)
            }
            Some(c) => {
                self.span.col_end += 1;
                self.span.byte_end += 1;
                Some(c)
            }
            None => None,
        }
    }

    fn next_if(&mut self, pred: impl Fn(char) -> bool) -> Option<char> {
        match self.peek() {
            Some(c) if pred(c) => self.next(),
            _ => None,
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.src.peek().copied()
    }

    fn add_token(&mut self, kind: TokenKind) {
        let span = self.span();
        self.tokens.push(Token { kind, span });
    }

    fn add_token_with_span(&mut self, kind: TokenKind, span: Span) {
        self.tokens.push(Token { kind, span });
    }

    fn span(&mut self) -> Span {
        let span = self.span;
        self.span = Span {
            row_start: span.row_end,
            col_start: span.col_end,
            byte_start: span.byte_end,
            ..span
        };
        span
    }

    fn lex_number(&mut self, c: char) {
        let mut number = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_digit()) {
            number.push(c);
        }
        let span = self.span();
        self.add_token_with_span(
            TokenKind::Number(number.parse().expect("invalid number")),
            span,
        );
    }

    fn lex_identifier(&mut self, c: char) {
        let mut identifier = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_alphanumeric() || c == '_') {
            identifier.push(c);
        }

        let kind = match identifier.as_str() {
            "push" => TokenKind::KeywordPush,
            "pop" => TokenKind::KeywordPop,
            "inc" => TokenKind::KeywordInc,
            "load" => TokenKind::KeywordLoad,
            "add" => TokenKind::KeywordAdd,
            "sub" => TokenKind::KeywordSub,
            "div" => TokenKind::KeywordDiv,
            "mul" => TokenKind::KeywordMul,
            "eq" => TokenKind::KeywordEq,
            "jne" => TokenKind::KeywordJne,
            "hult" => TokenKind::KeywordHult,
            "printreg" => TokenKind::KeywordPrintReg,
            "and" => TokenKind::KeywordAnd,
            "or" => TokenKind::KeywordOr,
            _ => TokenKind::Identifier(identifier),
        };
        let span = self.span();
        self.add_token_with_span(kind, span);
    }

    fn comment(&mut self) {
        while let Some(_) = self.next_if(|c| c != '\n') {}
        self.span();
    }

    fn lex(mut self) -> Vec<Token> {
        while let Some(c) = self.next() {
            match c {
                '0'..='9' => self.lex_number(c),
                'a'..='z' | 'A'..='Z' | '_' => self.lex_identifier(c),
                ' ' | '\t' => {
                    self.span();
                }
                ';' => self.comment(),
                '=' => self.add_token(TokenKind::Equal),
                '\n' => self.add_token(TokenKind::Delimiter),
                '(' => self.add_token(TokenKind::LeftParen),
                ')' => self.add_token(TokenKind::RightParen),
                '{' => self.add_token(TokenKind::LeftBrace),
                '}' => self.add_token(TokenKind::RightBrace),
                '[' => self.add_token(TokenKind::LeftBracket),
                ']' => self.add_token(TokenKind::RightBracket),
                ',' => self.add_token(TokenKind::Comma),
                '.' => self.add_token(TokenKind::Period),
                ':' => self.add_token(TokenKind::Colon),
                '%' => self.add_token(TokenKind::PercentSign),
                '&' => self.add_token(TokenKind::Ampersand),
                _ => panic!("Unexpected character: {}", c),
            }
        }
        self.tokens
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    pub fn snapshot_lexing(input: &str) -> String {
        let tokens = lex(input);
        let mut tokens = std::collections::VecDeque::from(tokens);
        let mut output = String::new();
        for (row, line) in input.lines().enumerate() {
            output += line;
            output += "\n";
            while let Some(tok) = tokens.pop_front() {
                if tok.span.row_end != row {
                    tokens.push_front(tok);
                    break;
                }
                output += &" ".repeat(tok.span.col_start);
                output += &"^".repeat(tok.span.col_end.saturating_sub(tok.span.col_start));
                output += &format!(" {tok:?}");
                output += "\n"
            }
        }
        output
    }

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_lexing(contents));
                });
            }
        };
    }

    snapshot!(demo, "../samples/demo.asm");
    snapshot!(main, "../samples/main.asm");
    snapshot!(fib, "../samples/fib.asm");
}
