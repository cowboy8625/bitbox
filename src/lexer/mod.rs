mod lexer;
pub mod token;

#[cfg(test)]
mod test;

pub fn lex(src: &str) -> Vec<Box<dyn token::Token>> {
    lexer::Lexer::new(src).collect()
}
