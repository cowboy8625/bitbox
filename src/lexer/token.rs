pub type Span = std::ops::Range<usize>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub span: Span,
}

impl Token {
    pub fn is_keyword(&self, keyword: Keyword) -> bool {
        matches!(self.kind, TokenKind::Keyword(ref k) if k == &keyword)
    }

    pub fn is_instruction(&self, instruction: Instruction) -> bool {
        matches!(self.kind, TokenKind::Instruction(ref i) if i == &instruction)
    }

    pub fn is_directive(&self, directive: Directive) -> bool {
        matches!(self.kind, TokenKind::Directive(ref d) if d == &directive)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum TokenKind {
    Colon,
    Comma,
    Directive(Directive),
    Dot,
    Equals,
    Identifier,
    Instruction(Instruction),
    InvalidToken,
    Keyword(Keyword),
    LeftBrace,
    LeftBracket,
    LeftParen,
    Number,
    PathSeparator,
    Plus,
    RightBrace,
    RightBracket,
    RightParen,
    Semicolon,
    Star,
    String,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Instruction {
    Add,
    Call,
    Phi,
    Ret,
    Sub,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Directive {
    Len,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Keyword {
    Const,
    Import,
    Function,
    Public,
}
