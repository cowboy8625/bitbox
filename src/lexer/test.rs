use crate::lexer::lex;
use std::fmt::Write;

pub fn snapshot_lexing(input: &str) -> String {
    let tokens = lex(input);
    let mut tokens = std::collections::VecDeque::from(tokens);
    let mut output = String::new();
    let mut total = 0;
    for line in input.lines() {
        output += line;
        output += "\n";
        while let Some(tok) = tokens.pop_front() {
            if total + line.len() <= tok.get_span().start {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.get_span().start - total);
            output += &"^".repeat(tok.get_span().len());
            write!(&mut output, " {tok:?}").expect("failed to write()");
            output += "\n"
        }

        total += line.len() + 1;
    }

    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(binary, "../../snapshots/basic.bitbox");
snapshot!(import_function, "../../snapshots/import_function.bitbox");
