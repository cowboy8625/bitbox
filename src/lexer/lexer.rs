use super::token;
use crate::ast::{
    BBString, Builtin, Colon, Comma, Dot, Equals, Identifier, InvalidToken, LeftBrace, LeftParen,
    Number, Plus, RightBrace, RightParen, Semicolon,
};

type Token = Box<dyn token::Token>;
pub struct Lexer<'a> {
    chars: std::iter::Peekable<std::str::Chars<'a>>,
    span: token::Span,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            chars: input.chars().peekable(),
            span: 0..0,
        }
    }

    fn next(&mut self) -> Option<char> {
        let value = self.chars.next();
        if value.is_some() {
            self.span.end += 1;
        }
        value
    }

    fn next_if(&mut self, predicate: impl Fn(char) -> bool) -> Option<char> {
        match self.chars.peek() {
            Some(value) if predicate(*value) => self.next(),
            _ => None,
        }
    }

    fn spanned(&mut self) -> token::Span {
        let span = self.span.clone();
        self.span = self.span.end..self.span.end;
        span
    }

    fn parse_number(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);

        while let Some(value) =
            self.next_if(|value| value.is_ascii_digit() || ['.', '_'].contains(&value))
        {
            lexeme.push(value);
        }

        token::create::<Number>(lexeme, self.spanned())
    }

    fn parse_identifier(&mut self, value: char) -> Token {
        let mut lexeme = String::from(value);
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Identifier>(lexeme, self.spanned())
    }

    fn parse_builtin(&mut self) -> Token {
        let mut lexeme = String::from('@');
        while let Some(value) = self.next_if(|value| value.is_ascii_alphanumeric() || value == '_')
        {
            lexeme.push(value);
        }

        token::create::<Builtin>(lexeme, self.spanned())
    }

    fn parse_string(&mut self) -> Token {
        let mut lexeme = String::from('#');
        while let Some(value) = self.next() {
            lexeme.push(value);
            if lexeme.ends_with("\"#") {
                break;
            }
        }
        token::create::<BBString>(lexeme, self.spanned())
    }

    fn parse(&mut self) -> Option<Token> {
        match self.next() {
            Some(value @ '0'..='9') => Some(self.parse_number(value)),
            Some(value) if value.is_ascii_alphabetic() => Some(self.parse_identifier(value)),
            Some(value) if value.is_ascii_whitespace() => {
                self.spanned();
                self.parse()
            }
            Some('#') if self.chars.peek() == Some(&'"') => Some(self.parse_string()),
            Some('@') => Some(self.parse_builtin()),
            Some('+') => Some(token::create::<Plus>('+', self.spanned())),
            Some('(') => Some(token::create::<LeftParen>('(', self.spanned())),
            Some(')') => Some(token::create::<RightParen>(')', self.spanned())),
            Some('{') => Some(token::create::<LeftBrace>('(', self.spanned())),
            Some('}') => Some(token::create::<RightBrace>(')', self.spanned())),
            Some(':') => Some(token::create::<Colon>(':', self.spanned())),
            Some(';') => Some(token::create::<Semicolon>(';', self.spanned())),
            Some(',') => Some(token::create::<Comma>(',', self.spanned())),
            Some('=') => Some(token::create::<Equals>('=', self.spanned())),
            Some('.') => Some(token::create::<Dot>('.', self.spanned())),
            Some(value) => Some(token::create::<InvalidToken>(value, self.span.clone())),
            None => None,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        self.parse()
    }
}

#[test]
fn test_lexer() {
    let token = Lexer::new("#\"test\"#").parse().unwrap();
    assert_eq!(token.get_lexeme(), "#\"test\"#");
}
