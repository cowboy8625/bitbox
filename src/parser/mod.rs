#[cfg(test)]
mod test;
use crate::ast::{self, Identifier};
use crate::lexer::token::{Span, Token};
use crate::ssa::{self, IntoSsaType};
use crate::stream::TokenStream;

#[derive(Debug)]
pub enum ParseError {
    UnexpectedToken {
        expected: String,
        found: String,
        span: Span,
    },
    UnexpectedEndOfStream,
}

enum TopLevel {
    Function(ssa::Function),
    Import(ssa::Import),
    Constant(ssa::Constant),
}

pub struct Parser {
    stream: TokenStream,
}

impl Parser {
    pub fn new(tokens: Vec<Box<dyn Token>>) -> Self {
        Self {
            stream: TokenStream::new(tokens),
        }
    }

    pub fn parse(&mut self) -> Result<ssa::Program, ParseError> {
        let mut imports = vec![];
        let mut functions = vec![];
        let mut constants = vec![];

        while self.stream.is_not_at_end() {
            match self.parse_top_level()? {
                TopLevel::Import(import) => imports.push(import),
                TopLevel::Function(func) => functions.push(func),
                TopLevel::Constant(constant) => constants.push(constant),
            }
        }

        Ok(ssa::Program {
            functions,
            imports,
            constants,
        })
    }
    fn parse_top_level(&mut self) -> Result<TopLevel, ParseError> {
        let visibility = self.parse_visibility();
        if self.peek_is_identifier("function") {
            Ok(TopLevel::Function(self.parse_function(visibility)?))
        } else if self.peek_is_identifier("import") {
            Ok(TopLevel::Import(self.parse_import()?))
        } else if self.peek_is_identifier("const") {
            Ok(TopLevel::Constant(self.parse_constant()?))
        } else {
            Err(ParseError::UnexpectedToken {
                expected: "top-level construct".to_string(),
                found: format!("{:?}", self.stream.peek_blind()),
                span: self.stream.current_span(),
            })
        }
    }

    fn parse_visibility(&mut self) -> ssa::Visibility {
        if self.peek_is_identifier("public") {
            self.consume::<ast::Identifier>().ok(); // Consume if present
            ssa::Visibility::Public
        } else {
            ssa::Visibility::Private
        }
    }

    fn consume<Expected>(&mut self) -> Result<Expected, ParseError>
    where
        Expected: Token + Clone,
    {
        match self.stream.next() {
            Some(value) if value.as_any().downcast_ref::<Expected>().is_some() => {
                Ok(value.as_any().downcast_ref::<Expected>().unwrap().clone())
            }
            Some(value) => Err(ParseError::UnexpectedToken {
                expected: std::any::type_name::<Expected>().to_string(),
                found: format!("{:?}", value),
                span: value.get_span(),
            }),
            None => Err(ParseError::UnexpectedEndOfStream),
        }
    }

    fn peek_is_identifier(&self, value: &str) -> bool {
        self.stream
            .peek::<ast::Identifier>()
            .map_or(false, |id| id.get_lexeme() == value)
    }

    fn consume_identifier(&mut self, value: &str) -> Result<ast::Identifier, ParseError> {
        let tok = self.consume::<Identifier>()?;
        if tok.lexeme != value {
            return Err(ParseError::UnexpectedToken {
                expected: value.to_string(),
                found: tok.lexeme.to_string(),
                span: tok.span,
            });
        }
        Ok(tok)
    }

    fn peek_is_builtin(&self, value: &str) -> bool {
        let Some(peek) = self.stream.peek::<ast::Builtin>().map(|i| i.get_lexeme()) else {
            return false;
        };

        value == peek
    }

    fn parse_function(&mut self, visibility: ssa::Visibility) -> Result<ssa::Function, ParseError> {
        self.consume_identifier("function")?;
        let func_name = self.consume::<ast::Identifier>()?;
        let params = self.parse_function_params()?;
        let return_type = self
            .consume::<ast::Identifier>()?
            .into_ssa_type()
            .map_err(|ident| ParseError::UnexpectedToken {
                expected: "Type".to_string(),
                found: ident.get_lexeme(),
                span: ident.get_span(),
            })?;
        let blocks = self.parse_function_block()?;

        Ok(ssa::Function {
            visibility,
            name: func_name.get_lexeme(),
            params,
            return_type,
            blocks,
        })
    }

    fn parse_function_params(&mut self) -> Result<Vec<ssa::Variable>, ParseError> {
        self.consume::<ast::LeftParen>()?;
        let mut params = vec![];

        let mut version = 0;
        while self.stream.is_not_at_end() && !self.stream.is_peek_a::<ast::RightParen>() {
            let name = self.consume::<ast::Identifier>()?;
            self.consume::<ast::Colon>()?;
            let ty = self.parse_type()?;

            let param = ssa::Variable { name, ty, version };
            params.push(param);
            if !self.stream.is_peek_a::<ast::Comma>() {
                break;
            }
            self.consume::<ast::Comma>()?;
            version += 1;
        }
        self.consume::<ast::RightParen>()?;
        Ok(params)
    }

    fn parse_function_block(&mut self) -> Result<Vec<ssa::BasicBlock>, ParseError> {
        let mut blocks = vec![];
        self.consume::<ast::LeftBrace>()?;
        while self.stream.is_not_at_end() && !self.stream.is_peek_a::<ast::RightBrace>() {
            let block = self.parse_basic_block()?;
            blocks.push(block);
        }
        self.consume::<ast::RightBrace>()?;
        Ok(blocks)
    }

    fn parse_basic_block(&mut self) -> Result<ssa::BasicBlock, ParseError> {
        let mut instructions = vec![];

        while self.stream.is_not_at_end() {
            if let Some(instruction) = self.parse_instruction()? {
                instructions.push(instruction);
            } else {
                break;
            }
        }

        Ok(ssa::BasicBlock {
            id: 0,
            instructions,
            // TODO: Add predecessors and successors
            successors: vec![],
            predecessors: vec![],
        })
    }

    fn parse_instruction(&mut self) -> Result<Option<ssa::Instruction>, ParseError> {
        if self.peek_is_builtin("@ret") {
            self.consume::<ast::Builtin>()?;
            let value = self.parse_operand()?;
            self.consume::<ast::Semicolon>()?;
            return Ok(Some(ssa::Instruction::Return(value)));
        }

        let name = self.consume::<ast::Identifier>()?;
        self.consume::<ast::Colon>()?;
        let ty = self.parse_type()?;
        self.consume::<ast::Equals>()?;

        if let Some(instruction) = self.parse_special_instruction(name.clone(), ty.clone())? {
            return Ok(Some(instruction));
        }

        Err(ParseError::UnexpectedToken {
            expected: "valid instruction".to_string(),
            found: name.lexeme,
            span: name.span,
        })
    }

    fn parse_special_instruction(
        &mut self,
        name: ast::Identifier,
        ty: ssa::Type,
    ) -> Result<Option<ssa::Instruction>, ParseError> {
        if self.peek_is_builtin("@add") {
            self.consume::<ast::Builtin>()?;
            let lhs = self.parse_operand()?;
            self.consume::<ast::Comma>()?;
            let rhs = self.parse_operand()?;
            self.consume::<ast::Semicolon>()?;
            return Ok(Some(ssa::Instruction::Add(
                ssa::Variable {
                    name,
                    ty,
                    version: 0,
                },
                lhs,
                rhs,
            )));
        }

        if self.peek_is_builtin("@sub") {
            self.consume::<ast::Builtin>()?;
            let lhs = self.parse_operand()?;
            self.consume::<ast::Comma>()?;
            let rhs = self.parse_operand()?;
            self.consume::<ast::Semicolon>()?;
            return Ok(Some(ssa::Instruction::Sub(
                ssa::Variable {
                    name,
                    ty,
                    version: 0,
                },
                lhs,
                rhs,
            )));
        }

        if self.peek_is_builtin("@call") {
            self.consume::<ast::Builtin>()?;
            let func_name = self.consume::<ast::Identifier>()?;
            let arguments = self.parse_arguments()?;
            self.consume::<ast::Semicolon>()?;
            return Ok(Some(ssa::Instruction::Call(
                ssa::Variable {
                    name,
                    ty,
                    version: 0,
                },
                func_name,
                arguments,
            )));
        }

        Ok(None)
    }

    fn parse_arguments(&mut self) -> Result<Vec<ssa::Operand>, ParseError> {
        self.consume::<ast::LeftParen>()?;
        let mut args = vec![];
        while self.stream.is_not_at_end() && !self.stream.is_peek_a::<ast::RightParen>() {
            args.push(self.parse_operand()?);
            if !self.stream.is_peek_a::<ast::Comma>() {
                break;
            }
            self.consume::<ast::Comma>()?;
        }
        self.consume::<ast::RightParen>()?;
        Ok(args)
    }

    fn parse_add_instruction(
        &mut self,
        name: ast::Identifier,
        ty: ssa::Type,
    ) -> Result<ssa::Instruction, ParseError> {
        self.consume::<ast::Builtin>()?;
        let var = ssa::Variable {
            name,
            ty,
            version: 0,
        };
        let lhs = self.parse_operand()?;
        self.consume::<ast::Comma>()?;
        let rhs = self.parse_operand()?;
        self.consume::<ast::Semicolon>()?;
        Ok(ssa::Instruction::Add(var, lhs, rhs))
    }

    fn parse_operand(&mut self) -> Result<ssa::Operand, ParseError> {
        if self.stream.is_peek_a::<ast::Number>() {
            let tok = self.consume::<ast::Number>()?;
            return Ok(ssa::Operand::Constant(tok));
        }
        let tok = self.consume::<ast::Identifier>()?;
        Ok(ssa::Operand::Variable(tok))
    }

    fn parse_import_function_params(&mut self) -> Result<Vec<ssa::Type>, ParseError> {
        let mut params = vec![];
        self.consume::<ast::LeftParen>()?;
        while self.stream.is_not_at_end() {
            let ty = self.parse_type()?;
            params.push(ty);
            if !self.stream.is_peek_a::<ast::Comma>() {
                break;
            }
            self.consume::<ast::Comma>()?;
        }
        self.consume::<ast::RightParen>()?;
        Ok(params)
    }

    fn parse_import(&mut self) -> Result<ssa::Import, ParseError> {
        self.consume_identifier("import")?;
        let tok = self.consume::<ast::Identifier>()?;
        match tok.get_lexeme().as_str() {
            "function" => {
                let module_name = self.consume::<ast::Identifier>()?;
                self.consume::<ast::PathSeparator>()?;
                let name = self.consume::<ast::Identifier>()?;
                let params = self.parse_import_function_params()?;
                let return_type =
                    self.consume::<ast::Identifier>()?
                        .into_ssa_type()
                        .map_err(|ident| ParseError::UnexpectedToken {
                            expected: "Type".to_string(),
                            found: ident.get_lexeme(),
                            span: ident.get_span(),
                        })?;
                self.consume::<ast::Semicolon>()?;
                Ok(ssa::Import::Function(ssa::FunctionSpec {
                    module_name,
                    name,
                    params,
                    return_type,
                }))
            }
            _ => Err(ParseError::UnexpectedToken {
                expected: "function".to_string(),
                found: tok.get_lexeme(),
                span: tok.get_span(),
            }),
        }
    }

    fn parse_constant_value(&mut self) -> Result<ssa::ConstantValue, ParseError> {
        if self.stream.is_peek_a::<ast::BBString>() {
            let tok = self.consume::<ast::BBString>()?;
            return Ok(ssa::ConstantValue::String(tok.lexeme.clone()));
        } else if self.stream.is_peek_a::<ast::Directive>() {
            let tok = self.consume::<ast::Directive>()?;
            let value = self.consume::<ast::Identifier>()?;
            let directive = match tok.get_lexeme().as_str() {
                ".len" => ssa::Directive::Len(value),
                _ => unimplemented!("unimplemented directive: {}", tok.get_lexeme()),
            };
            return Ok(ssa::ConstantValue::Directive(directive));
        }
        unimplemented!("{:?}", self.stream.peek_blind());
    }

    fn parse_constant(&mut self) -> Result<ssa::Constant, ParseError> {
        self.consume_identifier("const")?;
        let name = self.consume::<ast::Identifier>()?;
        self.consume::<ast::Colon>()?;
        let ty = self.parse_type()?;
        self.consume::<ast::Equals>()?;
        let value = self.parse_constant_value()?;
        self.consume::<ast::Semicolon>()?;
        Ok(ssa::Constant { name, ty, value })
    }

    fn parse_type(&mut self) -> Result<ssa::Type, ParseError> {
        if self.stream.is_peek_a::<ast::Identifier>() {
            let tok = self.consume::<ast::Identifier>()?;
            return tok
                .into_ssa_type()
                .map_err(|ident| ParseError::UnexpectedToken {
                    expected: "Type".to_string(),
                    found: ident.get_lexeme(),
                    span: ident.get_span(),
                });
        } else if self.stream.is_peek_a::<ast::Star>() {
            self.consume::<ast::Star>()?;
            let ty = self.parse_type()?;
            Ok(ssa::Type::Pointer(Box::new(ty)))
        } else if self.stream.is_peek_a::<ast::LeftBracket>() {
            self.consume::<ast::LeftBracket>()?;
            let count = self.consume::<ast::Number>()?;
            self.consume::<ast::Semicolon>()?;
            let ty = self.parse_type()?;
            self.consume::<ast::RightBracket>()?;
            Ok(ssa::Type::Array(
                count.get_lexeme().parse().unwrap(),
                Box::new(ty),
            ))
        } else {
            unreachable!("parse_type {:?}", self.stream.peek_blind())
        }
    }
}
