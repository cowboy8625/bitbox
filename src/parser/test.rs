use crate::lexer::lex;
use crate::parser;

pub fn snapshot_parsing(input: &str) -> String {
    let tokens = lex(input);
    let program = parser::Parser::new(tokens).parse();
    format!("{:#?}", program)
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_parsing(contents));
            });
        }
    };
}

snapshot!(binary, "../../snapshots/basic.bitbox");

