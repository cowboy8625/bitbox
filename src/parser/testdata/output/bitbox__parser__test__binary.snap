---
source: src/parser/test.rs
expression: snapshot_parsing(contents)
---
Ok(
    Program {
        imports: [],
        constants: [],
        functions: [
            Function {
                visibility: Public,
                name: "add",
                params: [
                    Variable {
                        name: Identifier {
                            lexeme: "x",
                            span: 20..21,
                        },
                        ty: Signed(
                            32,
                        ),
                        version: 0,
                    },
                    Variable {
                        name: Identifier {
                            lexeme: "y",
                            span: 28..29,
                        },
                        ty: Signed(
                            32,
                        ),
                        version: 1,
                    },
                ],
                return_type: Signed(
                    32,
                ),
                blocks: [
                    BasicBlock {
                        id: 0,
                        instructions: [
                            Add(
                                Variable {
                                    name: Identifier {
                                        lexeme: "z",
                                        span: 46..47,
                                    },
                                    ty: Signed(
                                        32,
                                    ),
                                    version: 0,
                                },
                                Variable(
                                    Identifier {
                                        lexeme: "x",
                                        span: 61..62,
                                    },
                                ),
                                Variable(
                                    Identifier {
                                        lexeme: "y",
                                        span: 64..65,
                                    },
                                ),
                            ),
                            Return(
                                Variable(
                                    Identifier {
                                        lexeme: "z",
                                        span: 76..77,
                                    },
                                ),
                            ),
                        ],
                        successors: [],
                        predecessors: [],
                    },
                ],
            },
        ],
    },
)
